sonar-dotnet-shared-library/src/main/java/org/sonarsource/dotnet/shared/plugins/FileTypeSensor.java

```diff
@@ -0,0 +1,95 @@
+
+/**
+ * This class is a non-global sensor used to count the type of files in the .NET projects (i.e. Scanner modules).
+ *
+ * Why is this needed?
+ * - the Scanner for MSBuild categorizes projects as MAIN or TEST (see https://github.com/SonarSource/sonar-scanner-msbuild/wiki/Analysis-of-product-projects-vs.-test-projects)
+ * - in SQ / SC, users can specify which files should be considered as MAIN (sources) or TEST (test sources) (see https://docs.sonarqube.org/latest/project-administration/narrowing-the-focus/)
+ * - the categorization is not obvious, so this additional information should help users debug when needed
+ */
+@ScannerSide
+public class FileTypeSensor implements Sensor {
+  private static final Logger LOG = Loggers.get(FileTypeSensor.class);
+
+  private final ProjectTypeCollector projectTypeCollector;
+  private final DotNetPluginMetadata pluginMetadata;
+
+  public FileTypeSensor(ProjectTypeCollector projectTypeCollector, DotNetPluginMetadata pluginMetadata) {
+    this.projectTypeCollector = projectTypeCollector;
+    this.pluginMetadata = pluginMetadata;
+  }
+
+  @Override
+  public void describe(SensorDescriptor descriptor) {
+    String name = String.format("Verify what types of files (MAIN, TEST) are in %s projects.", pluginMetadata.languageKey());
+    descriptor.name(name);
+    // we do not filter by language because we want to be called on projects without sources
+    // (that could reference only shared sources e.g. in .projitems)
+  }
+
+  @Override
+  public void execute(SensorContext context) {
+    FileSystem fs = context.fileSystem();
+    Configuration configuration = context.config();
+
+    boolean hasMainFiles = hasFilesOfType(fs, Type.MAIN);
+    boolean hasTestFiles = hasFilesOfType(fs, Type.TEST);
+
+    Optional<String> projectName = configuration.get("sonar.projectName");
+    if (projectName.isPresent()) {
+      LOG.debug("Adding file type information (has MAIN '{}', has TEST '{}') for project '{}' (base dir '{}'). For debug info, see ProjectInfo.xml in '{}'.",
+        hasMainFiles, hasTestFiles, projectName.get(), getValueOrEmpty(configuration.get("sonar.projectBaseDir")), getValueOrEmpty(getProjectOutPath(configuration)));
+      projectTypeCollector.addProjectInfo(hasMainFiles, hasTestFiles);
+    }
+  }
+
+  // this is duplicated from future DotNetSensor
+  private boolean hasFilesOfType(FileSystem fs, Type fileType) {
+    return fs.inputFiles(fs.predicates().and(fs.predicates().hasType(fileType), fs.predicates().hasLanguage(pluginMetadata.languageKey()))).iterator().hasNext();
+  }
+
+  private Optional<String> getProjectOutPath(Configuration configuration) {
+    if (pluginMetadata.languageKey().equalsIgnoreCase("cs")) {
+      return configuration.get("sonar.cs.analyzer.projectOutPath");
+    } else if (pluginMetadata.languageKey().equalsIgnoreCase("vbnet")) {
+      return configuration.get("sonar.vbnet.analyzer.projectOutPath");
+    } else {
+      return Optional.empty();
+    }
+  }
```

Comment:
There must be a better, generic way of doing this. In worst case scenario,
`return configuration.get("sonar."+pluginMetadata.languageKey().toLowerCase()+".analyzer.projectOutPath");`

Shared base should not be aware of explicit list of it's users.

Reply:
While looking at this, it seems save to assume we can use language key directly and don't care about case sensitivity issue.

 https://github.com/SonarSource/sonar-dotnet/blob/fb6f4ae1deda8d2e9fe0c8b8699744c596cea024/sonar-dotnet-shared-library/src/main/java/org/sonarsource/dotnet/shared/plugins/CodeCoverageProvider.java#L45-L62